### DISCUSSION TITLE
Upgrade to 3.10 on Raspberry Pi

### DISCUSSION THREAD

**Discussion by BentWrench** [jamulussoftware/jamulus#3183] [Technical Support & Help]
Hello All,

I noticed 3.10 is out and so I upgraded using my usual method. I usually compile from source, as I'm not aware of a Jamulus Raspberry Pi headless server I can install. I'm not a Linux expert by any means, so please take everything I say with a grain of salt (with skepticism).

I decided to see if anything had changed since version 3.9.1, so I tried the Instructions on the Jamulus.io which referred to the Linux instructions, and they don't work, at least not on Raspberry Pi which is based on Debian. Maybe there needs to be a separate page just for Pi users. Running Pi servers works great and is cheap.

Here are the instructions I use to update to the latest version, in case anyone out there needs them:
```
rm -r -f jamulus # delete existing jamulus folder
git clone https://github.com/jamulussoftware/jamulus.git
cd jamulus
git checkout # or git checkout r3_X_Y, where X and Y are the latest [official version numbers
qmake "CONFIG+=nosound headless" Jamulus.pro
sudo make clean
sudo systemctl stop jamulus
Sudo make install
sudo systemctl start jamulus
```

Cheers,

  -Rob

**Comment by rdica** [comment]
@BentWrench You should mention what version of the pi (cpu arch etc), Linux distro and version used.

Also, can you detail what didn't work when you said you tried the Instructions on the Jamulus.io which referred to the Linux instructions.

**Reply by BentWrench** [reply]
Hello rdica,

I have a "Raspberry Pi 4 Model B Rev 1.2" running "Raspbian GNU/Linux 10 (buster)".

I was going through the instructions to compile a headless server on this page: https://github.com/jamulussoftware/jamulus/blob/main/COMPILING.md

Under the "Download sources" section, I tried the git command and it did not work. when I change to the command to git clone `https://github.com/jamulussoftware/jamulus.git` then it works properly and downloads the source. Also it helps to remove the jamulus directory first using  `rm -r -f jamulus` 

Next, the dependency install works as written - no issue

After that, moved the "Headless" server build section, and ran the command `make distclean` and get this:
'make: *** No rule to make target 'distclean'.  Stop.' I think running this is optional and only necessary when you have previous files to clean up. I get around that by using rm as noted above.

I then ran `make clean` and then `qmake "CONFIG+=headless serveronly"` and that ran with no issues.

I then started the server with `Jamulus --nogui --server --directoryaddress anygenre2.jamulus.io:22224 --serverinfo "Rob's Test Pi;Any Town;us"` and that worked fine.

You can't use the jamulus-headless command as the default name of the binary is Jamulus.

Next I went down the rabbit hole to setup a new server using the compile option on my extra Pi. You have to create a new unit file from scratch, and the instructions for doing that are not in the server manual, or they've been moved. I set it up to use the jamulus-headless naming convention like the automated install from the repo does.

Here is what I came up with:

#new headless install for Raspberry Pi using the compile option
```
git clone https://github.com/jamulussoftware/jamulus.git
cd jamulus
qmake "CONFIG+=headless serveronly"
make clean
sudo make
sudo cp Jamulus /usr/local/bin/jamulus-headless # copy the new Jamulus binary to bin folder as jamulus-headless
sudo adduser --system --no-create-home jamulus #create a jamulus user for the service
sudo systemctl edit --force --full jamulus-headless # create new service file
```
Paste these into the service file in Nano
Beginning of file contents - don't include this line
```
[Unit]
Description=Jamulus-Server
After=network.target

[Service]
Type=simple
User=jamulus
Group=nogroup
NoNewPrivileges=true
ProtectSystem=true
ProtectHome=true
Nice=-20
IOSchedulingClass=realtime
IOSchedulingPriority=0

#### Change this ####
ExecStart=/usr/local/bin/jamulus-headless -s -n -e anygenre2.jamulus.io:22224 -o "Rob's Test Pi;AnytownYouWant;us"

Restart=on-failure

[Install]
WantedBy=multi-user.target
```

End of service file contents - don't include this line
```
sudo systemctl daemon-reload # load all the service files
sudo systemctl start jamulus-headless # start jamulus-headless to make sure it works
sudo systemctl enable jamulus-headless # enable the service to start at boot

```

Anyway...maybe too much info or someone has already done this with a script, but I had some extra time to work on this so here ya go.

**Reply by rdica** [reply]
Thanks @BentWrench I'm sure there are those who will find this helpful when attempting the build on this platform.
I build on Fedora and Ubuntu on home and various cloud-provided servers so I'm always pleased to see someone attemptiing builds!

**Comment by ann0see** [comment]
If you're using raspbian, you should be able to just add the repository and install jamulus-headless as described on the Linux install page. We do provide a armhf headless build

**Reply by BentWrench** [reply]
Hello ann0see,
I gave this a try, using the instructions on this page: https://jamulus.io/wiki/Running-a-Server. I'm running "Raspbian GNU/Linux 10 (buster)"

When I get to this command: `sudo ./setup_repo.sh && sudo apt install jamulus-headless`
I get errors these errors:

```
Setting up Jamulus repo at /etc/apt/sources.list.d/jamulus.list...
Installing Jamulus GPG key at /etc/apt/trusted.gpg.d/jamulus.asc...
Running apt update...
E: Failed to fetch https://github.com/jamulussoftware/jamulus/releases/latest/download/./InRelease  401  Unauthorized [IP: 185.199.108.133 443]
E: The repository 'https://github.com/jamulussoftware/jamulus/releases/latest/download ./ InRelease' is not signed.
You should now be able to install a full Jamulus package via
  apt install jamulus
or a server-only, dependency-reduced build via
  apt install jamulus-headless

This package will automatically be updated when you perform system updates.
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package jamulus-headless
```

**Reply by ann0see** [reply]
I believe that's due to an old version of apt which doesn't follow redirects. Can you update your OS?

**Reply by ann0see** [reply]
See https://github.com/orgs/jamulussoftware/discussions/3180

**Reply by BentWrench** [reply]
You are correct, it was a problem with apt and Raspbian Buster that's been fixed in Raspberry Pi OS (AKA Raspbian Bullseye). I installed the lite version on a spare SD card and tested it out. The install worked correctly, with one issue. The headless server would start then quit. Some troubleshooting lead me to this article: https://github.com/jamulussoftware/jamulus/issues/1515

When I edited the service unit file, I changed the start-up line:
It went from `ExecStart=/bin/sh -c exec '/usr/bin/jamulus-headless -s -n'`
To `ExecStart=/usr/bin/jamulus-headless -s -n -e anygenre2.jamulus.io:22224 -o "Rob's Test Pi;Portland OR;us"`

And that fixed the probelm, the service would start correctly. That article I linked explains the issue with the sh -c entry.

Thanks for the guidance!

**Reply by ann0see** [reply]
Thanks for confirming. 
I think we might want to change the file https://github.com/jamulussoftware/jamulus/blob/main/linux/debian/jamulus-headless.service to maki installation easier.

**Reply by BentWrench** [reply]
Good idea. How do we go about changing the file?

**Reply by mcfnord** [reply]
@ann0see has created [an issue about changing this](https://github.com/jamulussoftware/jamulus/issues/3185). ROB'S PI!

**Reply by ann0see** [reply]
Feel free to edit the file yourself and open a PR.

**Reply by rdica** [reply]
PR submitted.

**Reply by mcfnord** [reply]
https://github.com/jamulussoftware/jamulus/pull/3188

**Reply by mcfnord** [reply]
oh man! you beat me to it!

**Comment by BentWrench** [comment]
> PR submitted.

Thanks! I've never done one before so glad you submitted it.
